### PR TITLE
Adds validation to check user belongs to linked LMS

### PR DESCRIPTION
**Pull request by rehan99000**
**Description:**
Adds a validation that checks if the course being created belongs to the logged-in user.

**JIRA:**
https://edlyio.atlassian.net/browse/EDLY-1281

### CODE REVIEW

File: ecommerce/extensions/edly_ecommerce_app/helpers.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    Arguments:
+        encoded_cookie_data (dict): Edly user info cookie JWT encoded string.
+
+    Returns:
+        dict
+    """
+    return jwt.decode(encoded_cookie_data, settings.EDLY_COOKIE_SECRET_KEY, algorithms=[settings.EDLY_JWT_ALGORITHM])
+
+def validate_course_id(course_id, request):
```
Comment (ziafazal) [inline ecommerce/extensions/edly_ecommerce_app/helpers.py:18]: can we rename this method to `valid_course_id`

File: ecommerce/extensions/edly_ecommerce_app/helpers.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import jwt
+from django.conf import settings
```
Comment (zubair-arbi) [inline ecommerce/extensions/edly_ecommerce_app/helpers.py:2]: nit: fix imports ordering
https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#imports-order

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 1 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:18]: @ziafazal @rehan99000 how about changing name to `is_valid_edly_site_course`?

@rehan99000 also there should be two new lines before this method

File: ecommerce/extensions/edly_ecommerce_app/helpers.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        course_id (string): Course ID string.
+        request (WSGI Request): Django Request object.
+
+    Returns:
+        boolean
+    """
+    partner_short_code = request.site.partner.short_code
+    course_key = CourseKey.from_string(course_id)
+    edly_user_info_cookie = request.COOKIES.get(settings.EDLY_USER_INFO_COOKIE_NAME, None)
+    edx_org_short_name = decode_edly_user_info_cookie(edly_user_info_cookie).get('edx-org', None)
```
Comment (zubair-arbi) [inline ecommerce/extensions/edly_ecommerce_app/helpers.py:32]: Add a method `get_edx_org_from_edly_cookie` (https://github.com/edly-io/edx-platform/pull/73/files)
```python

def get_edly_sub_org_from_cookie(encoded_cookie_data):
    """
    Returns "edly-sub-org" value from the "edly-user-info" cookie.

    Arguments:
        encoded_cookie_data (dict): Edly user info cookie JWT encoded string.

    Returns:
        string
    """

    if not encoded_cookie_data:
        return ''

    decoded_cookie_data = decode_edly_user_info_cookie(encoded_cookie_data)
    return decoded_cookie_data['edly-sub-org']
```

File: ecommerce/extensions/edly_ecommerce_app/helpers.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    # We assume that the ECOM site partner short code will always be the same as short name of its
```
Comment (zubair-arbi) [inline ecommerce/extensions/edly_ecommerce_app/helpers.py:33]: nit: usually the docstring and comment shouldn't exceed 80 characters length for readability:
```python
# We assume that the "short_code" value of ECOM site partner will always 
# be the same as "short_name" value of its related edx organization in LMS
```

File: ecommerce/extensions/edly_ecommerce_app/helpers.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    # related edx organization in LMS
+    if edx_org_short_name and course_key.org == edx_org_short_name and course_key.org == partner_short_code:
```
Comment (zubair-arbi) [inline ecommerce/extensions/edly_ecommerce_app/helpers.py:35]:
```python
if course_key.org == edx_org_short_name and course_key.org == partner_short_code:
```

File: ecommerce/extensions/api/v2/views/courses.py [hunk 6]
```diff
@@ -71,6 +72,11 @@ def retrieve(self, request, *args, **kwargs):
               paramType: query
               multiple: false
         """
+        course_id = kwargs.get('pk')
+        if not validate_course_id(course_id, request):
+            return Response(
```
Comment (zubair-arbi) [inline ecommerce/extensions/api/v2/views/courses.py:77]: Always use proper code formatting.
```python
return Response(
    {
        'error': 'Course with ID "{course_id}" does not exist for this site'.format(course_id=course_id)
    },
   status=status.HTTP_404_NOT_FOUND
)
```

File: ecommerce/extensions/api/serializers.py [hunk 7]
```diff
@@ -518,6 +519,18 @@ def validate_products(self, products):
 
         return products
 
+    def validate_id(self, course_id):
+        """
+        Validate if given course ID belongs to logged in user's organization.
+        """
+        if not validate_course_id(course_id, self.context['request']):
+            raise serializers.ValidationError(
+                _(u'Course with ID "{course_id}" does not exist.').format(course_id=course_id)
```
Comment (zubair-arbi) [inline ecommerce/extensions/api/serializers.py:528]: Add test for this validation error

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 5 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:70]: there is a possibility of `edx_org_short_name` of being None so I have added that check

Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:70]: If its not set then that means the sites (LMS, ECOM) configurations for this client are not correct.

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 1 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: I have added two lines. and changed name according to @zubair-arbi suggestion

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 2 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:1]: fixed. I was assuming jwt is local. I have fixed it now

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 3 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: done

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 4 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: done

File: ecommerce/extensions/api/v2/views/courses.py (hunk 6 above)
Comment (rehan99000) [reply ecommerce/extensions/api/v2/views/courses.py:78]: done

File: ecommerce/extensions/api/serializers.py (hunk 7 above)
Comment (rehan99000) [reply ecommerce/extensions/api/serializers.py:528]: added

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 1 above)
Comment (ziafazal) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: @zubair-arbi we should avoid adding `edly` to any public repos. name of organization should never be used in code which is going to be public. we can use it in app name to differentiate things.

File: ecommerce/extensions/edly_ecommerce_app/tests/test_helpers.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        assert self.test_edly_user_info_cookie_data == decoded_edly_user_info_cookie_data
+
+    def test_get_edly_sub_org_from_cookie(self):
+        """
+        Test that "get_edx_org_from_cookie" method returns edx-org short name correctly.
+        """
+        edx_org_short_name = self.test_edly_user_info_cookie_data.get('edx-org')
+        edly_user_info_cookie_data = encode_edly_user_info_cookie(self.test_edly_user_info_cookie_data)
+        assert edx_org_short_name == get_edx_org_from_edly_cookie(edly_user_info_cookie_data)
+
```
Comment (rehan99000) [inline ecommerce/extensions/edly_ecommerce_app/tests/test_helpers.py:66]: These three tests have been copied from edly features app in openedx

File: ecommerce/static/js/views/form_view.js [hunk 9]
```diff
@@ -159,19 +162,21 @@ define([
...
                         message = response.responseJSON.error;
+                        AlertUtils.renderAlert('danger', gettext('Error!'), message, self);
 
                         // Log the error to the console for debugging purposes
                         console.error(message); // eslint-disable-line no-console
                     } else {
                         // Log the error to the console for debugging purposes
+                        for (var key in response.responseJSON) {
+                            AlertUtils.renderAlert('danger', gettext('Error!'), response.responseJSON[key], self);
+                        }
```
Comment (rehan99000) [inline ecommerce/static/js/views/form_view.js:177]: This renders alerts for any serializer raised validation errors

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 2 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:1]: Still needs to fix  openedx/features/edly/tests/test_utils.py
```python
import jwt
from mock import MagicMock

from django.conf import settings
from django.http import HttpResponse
from django.test import TestCase
from django.test.client import RequestFactory

from openedx.features.edly import cookies as cookies_api
from openedx.features.edly.tests.factories import EdlySubOrganizationFactory, EdlyUserProfileFactory, SiteFactory
from openedx.features.edly.utils import (
    create_user_link_with_edly_sub_organization,
    decode_edly_user_info_cookie,
    encode_edly_user_info_cookie,
    get_edly_sub_org_from_cookie,
    user_has_edly_organization_access
)
from student.tests.factories import UserFactory

```

File: ecommerce/extensions/api/v2/views/courses.py (hunk 6 above)
Comment (zubair-arbi) [reply ecommerce/extensions/api/v2/views/courses.py:78]: message still need to update

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 3 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: I am not able to see this method

File: ecommerce/static/js/views/form_view.js (hunk 9 above)
Comment (zubair-arbi) [reply ecommerce/static/js/views/form_view.js:177]: Got it

File: ecommerce/extensions/edly_ecommerce_app/helpers.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        cookie_data (dict): Edly user info cookie dict.
+
+    Returns:
+        string
+    """
+    return jwt.encode(cookie_data, settings.EDLY_COOKIE_SECRET_KEY, algorithm=settings.EDLY_JWT_ALGORITHM)
+
+
+def get_edx_org_from_edly_cookie(encoded_cookie_data):
+    """
```
Comment (rehan99000) [inline ecommerce/extensions/edly_ecommerce_app/helpers.py:35]: @zubair-arbi this is the `get_edx_or_from_edly_cookie` method

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 5 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:70]: Yes so we need to return `False` in that case. That's why there is the check

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 10 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:35]: Thanks

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 3 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: Found the method `get_edly_sub_org_from_cookie`  https://github.com/edly-io/ecommerce/pull/15/files#r427291130

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 5 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:70]: Ok fine. :)

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 1 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: @ziafazal got it.
@rehan99000 lets rename this method to `is_valid_site_course`

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 2 above)
Comment (zubair-arbi) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py:1]: Please ignore this. The imports are fine.

File: ecommerce/extensions/edly_ecommerce_app/helpers.py (hunk 1 above)
Comment (rehan99000) [reply ecommerce/extensions/edly_ecommerce_app/helpers.py]: okay

**Comment by zubair-arbi** [comment]
@rehan99000 Please merge this PR.

**Pull request merged**